events/mbed_events.cpp

```diff
@@ -0,0 +1,75 @@
+/* events
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "events/mbed_events.h"
+#include "mbed.h"
+
+using namespace events;
+
+namespace mbed {
+
+#ifdef MBED_CONF_RTOS_PRESENT
+/* Create an event queue, and start the thread that dispatches it. Static
+ * variables mean this happens once the first time each template instantiation
+ * is called. This is currently instantiated no more than twice.
+ */
+template
+<osPriority Priority, size_t QueueSize, size_t StackSize>
+EventQueue *do_shared_event_queue_with_thread()
+{
+    static unsigned char queue_buffer[QueueSize];
+    static EventQueue queue(sizeof queue_buffer, queue_buffer);
+
+    static unsigned char stack[StackSize];
+    static Thread thread(Priority, StackSize, stack);
+
+    Thread::State state = thread.get_state();
+    if (state == Thread::Inactive || state == Thread::Deleted) {
+        osStatus status = thread.start(callback(&queue, &EventQueue::dispatch_forever));
+
+        if (status != osOK) {
+            error("thread start status %d", status);
+            MBED_ASSERT(status == osOK);
+            return NULL;
+        }
+    }
+
+    return &queue;
+}
+#endif
+
+EventQueue *get_shared_normal_event_queue()
+{
+#if MBED_CONF_EVENTS_SHARED_NORMAL_DISPATCH_FROM_APPLICATION || !defined MBED_CONF_RTOS_PRESENT
+    /* Only create the EventQueue, but no dispatching thread */
+    static unsigned char queue_buffer[MBED_CONF_EVENTS_SHARED_NORMAL_EVENTSIZE];
+    static EventQueue queue(sizeof queue_buffer, queue_buffer);
+
+    return &queue;
+#else
+    return do_shared_event_queue_with_thread<osPriorityNormal, MBED_CONF_EVENTS_SHARED_NORMAL_EVENTSIZE, MBED_CONF_EVENTS_SHARED_NORMAL_STACKSIZE>();
+#endif
+}
+
+#ifdef MBED_CONF_RTOS_PRESENT
+EventQueue *get_shared_interrupt_deferral_event_queue()
+{
+    return do_shared_event_queue_with_thread<osPriorityRealtime, MBED_CONF_EVENTS_SHARED_IRQDEFERRAL_EVENTSIZE, MBED_CONF_EVENTS_SHARED_IRQDEFERRAL_STACKSIZE>();
```

Comment:
After responding below, I started thinking, rather than re-naming it (or in addition to renaming it), why not assert the latency instead? That would enforce the realtime nature of this queue. In addition, for mbed 2 builds you could share the event queue, but with caveat that adding long running events to the normal event queue while using a driver which requires the interrupt queue will trigger timing asserts. That way you could use components like nanostack on 2 as long as the application and any libraries it pulls in don't need to post any long-running tasks to the queue.